src/js/containers/RPEInput.tsx

```diff
@@ -0,0 +1,36 @@
+import * as React from "react";
+
+export type Validator<T> = (value: T) => string | null;
+export type ChangeHandler<T> = (id: string, value: T) => void;
+
+export interface RPEInputProps<T> {
+    id: string;
+    label: string;
+    value: T;
+    validator: Validator<T>;
+    onChange: ChangeHandler<T>;
+
+    className?: string;
+
+    inputProps?: React.InputHTMLAttributes<HTMLInputElement>;
+}
```

Comment:
So in standard react using typescript, does this take the place of proptypes validations?

Reply:
Exactly!! It's quite nice. The interfaces are used in generic type parameters to the class itself.